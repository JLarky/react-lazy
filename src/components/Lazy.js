import React from 'react'
import PropTypes from 'prop-types'

import Observer from '@researchgate/react-intersection-observer'

import { propsWithNoScriptRender } from '../lib/wrap'

class Lazy extends React.PureComponent {
    constructor(props) {
        super(props)

        this.state = { show: false }

        this.handleViewport = this.handleViewport.bind(this)
    }

    handleViewport(event, unobserve) {
        if (!event.isIntersecting || !this.props.visible) {
            return
        }

        if (this.props.onViewport) {
            this.props.onViewport(event)
        }

        if (event.defaultPrevented) {
            return
        }

        unobserve()

        this.setState({ show: true }, this.props.onLoad)
    }

    render() {
        const {
            children,
            clientOnly,
            component,
            cushion,
            forwardedRef: ref,
            ltIE9,
            visible,
            onLoad,
            onViewport,
            threshold,
            viewport,
            ...rest
        } = this.props

        const props = Object.assign({ ref }, rest)
        const isClientRender = clientOnly || this.state.show

        return (
            <Observer onChange={this.handleViewport} root={viewport} rootMargin={cushion} threshold={threshold}>
                {React.createElement(
                    component,
                    isClientRender ? props : propsWithNoScriptRender(children, ltIE9, props),
                    isClientRender && this.state.show && visible ? children : null,
                )}
            </Observer>
        )
    }
}

Lazy.defaultProps = {
    clientOnly: false,
    component: 'div',
    ltIE9: false,
    visible: true,
}

Lazy.propTypes = {
    clientOnly: PropTypes.bool,
    children: PropTypes.node,
    component: PropTypes.any,
    cushion: PropTypes.string,
    forwardedRef: PropTypes.any,
    ltIE9: PropTypes.bool,
    onLoad: PropTypes.func,
    onViewport: PropTypes.func,
    threshold: PropTypes.oneOfType([PropTypes.number, PropTypes.arrayOf(PropTypes.number)]),
    viewport: PropTypes.oneOfType(
        [PropTypes.string].concat(typeof HTMLElement === 'undefined' ? [] : PropTypes.instanceOf(HTMLElement)),
    ),
    visible: PropTypes.bool,
}

// eslint-disable-next-line react/display-name,react/no-multi-comp
export default React.forwardRef((props, ref) => <Lazy {...props} forwardedRef={ref} />)
