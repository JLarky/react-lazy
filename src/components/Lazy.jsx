import React from 'react'
import { findDOMNode } from 'react-dom'
import PropTypes from 'prop-types'

import { addElement, removeElement } from '../lib/viewport'
import { propsWithNoScriptRender } from '../lib/wrap'

class Lazy extends React.PureComponent {
    constructor(props) {
        super(props)

        this.state = { loadedAt: null }

        this.onViewport = this.onViewport.bind(this)
    }

    componentDidMount() {
        this.options = {
            callback: this.onViewport,
            cushion: this.props.cushion,
            element: findDOMNode(this),
        }
        addElement(this.options)
    }

    componentWillReceiveProps(nextProps) {
        if (nextProps.cushion !== this.props.cushion) {
            this.options.cushion = nextProps.cushion
        }
    }

    componentWillUnmount() {
        removeElement(this.options)
        delete this.options
    }

    onViewport() {
        const { onLoad, onViewport, visible } = this.props

        if (!visible) {
            return false
        }

        if (onViewport) {
            onViewport()
        }

        this.setState({ loadedAt: Date.now() }, onLoad)
    }

    render() {
        const { children, component, ltIE9, visible, ...props } = this.props
        // remove props we do not want to pass to the rendering component
        delete props.cushion
        delete props.onLoad
        delete props.onViewport

        if (visible && this.state.loadedAt) {
            return React.createElement(component, props, children)
        }

        // wrap all contents inside noscript
        return React.createElement(component, propsWithNoScriptRender(children, ltIE9, props))
    }
}

Lazy.defaultProps = {
    component: 'div',
    cushion: 0,
    ltIE9: false,
    visible: true,
}

const ReactElement = PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object,
    PropTypes.func,
])

Lazy.propTypes = {
    children: PropTypes.node,
    component: ReactElement,
    cushion: PropTypes.number,
    ltIE9: PropTypes.bool,
    onLoad: PropTypes.func,
    onViewport: PropTypes.func,
    visible: PropTypes.bool,
}

export default Lazy
