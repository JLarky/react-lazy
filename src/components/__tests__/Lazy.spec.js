/* eslint-env jest */
/* eslint-disable react/prop-types,react/no-multi-comp */
import React from 'react'
import renderer from 'react-test-renderer'

import Lazy from '../Lazy'

const originalIntersectionObserver = window.IntersectionObserver

function noop() {}

function mockIntersectionObserver() {
    return {
        disconnect: noop,
        observe: noop,
        unobserve: noop,
    }
}

const target = { nodeType: 1 }

beforeAll(() => {
    window.IntersectionObserver = mockIntersectionObserver
})

afterAll(() => {
    window.IntersectionObserver = originalIntersectionObserver
})

describe('Lazy', () => {
    test('should render as div and with empty noscript by default', () => {
        const tree = renderer.create(<Lazy />)
        expect(tree.toJSON()).toMatchSnapshot()
    })

    test('should let change component to be rendered', () => {
        const tree = renderer.create(<Lazy component="section" />)
        expect(tree.toJSON()).toMatchSnapshot()
    })

    test('ltIE9 should render with IE conditional comments around noscript', () => {
        const tree = renderer.create(<Lazy ltIE9 />)
        expect(tree.toJSON()).toMatchSnapshot()
    })

    test('children should be rendered inside noscript', () => {
        const tree = renderer.create(
            <Lazy>
                <span>I am inside noscript!</span>
            </Lazy>,
        )
        expect(tree.toJSON()).toMatchSnapshot()
    })

    test('with ltIE9 children should be rendered inside noscript with conditional comments', () => {
        const tree = renderer.create(
            <Lazy ltIE9>
                <span>I am inside IECC noscript!</span>
            </Lazy>,
        )
        expect(tree.toJSON()).toMatchSnapshot()
    })

    test('should render only div in clientOnly mode', () => {
        const tree = renderer.create(<Lazy clientOnly />)
        expect(tree.toJSON()).toMatchSnapshot()
    })

    test('should call ref callback', () => {
        const spy = jest.fn()
        renderer.create(<Lazy ref={spy} />, { createNodeMock: () => target })
        expect(spy).toHaveBeenCalledWith(target)
    })
})
