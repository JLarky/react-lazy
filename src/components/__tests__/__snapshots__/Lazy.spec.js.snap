// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lazy children should be rendered inside noscript 1`] = `
<div
  dangerouslySetInnerHTML={
    Object {
      "__html": "<noscript><span>I am inside noscript!</span></noscript>",
    }
  }
/>
`;

exports[`Lazy ltIE9 should render with IE conditional comments around noscript 1`] = `
<div
  dangerouslySetInnerHTML={
    Object {
      "__html": "<!--[if IE 9]><!--><noscript><!--<![endif]--><!--[if IE 9]><!--></noscript><!--<![endif]-->",
    }
  }
/>
`;

exports[`Lazy should let change component to be rendered 1`] = `
<section
  dangerouslySetInnerHTML={
    Object {
      "__html": "<noscript></noscript>",
    }
  }
/>
`;

exports[`Lazy should render as div and with empty noscript by default 1`] = `
<div
  dangerouslySetInnerHTML={
    Object {
      "__html": "<noscript></noscript>",
    }
  }
/>
`;

exports[`Lazy should render only div in clientOnly mode 1`] = `<div />`;

exports[`Lazy with ltIE9 children should be rendered inside noscript with conditional comments 1`] = `
<div
  dangerouslySetInnerHTML={
    Object {
      "__html": "<!--[if IE 9]><!--><noscript><!--<![endif]--><span>I am inside IECC noscript!</span><!--[if IE 9]><!--></noscript><!--<![endif]-->",
    }
  }
/>
`;
